### ISSUE TITLE
Self host ReadTheDocs docs building (was a problem with ReadTheDocs search)

### ISSUE THREAD

**Issue by tumd** [debops/debops#954]
Hi,

Searches on [docs.debops.org](https://docs.debops.org) that have results pointing to a defaults-configuration no longer work.

Example;
https://docs.debops.org/en/master/search.html?q=haproxy
The result of above search gives one result;
http://docs.debops.org/en/master/ansible/roles/debops.pki/defaults.html?highlight=haproxy
Which does not work. It should instead be;
http://docs.debops.org/en/master/ansible/roles/debops.pki/defaults/main.html?highlight=haproxy

I guess this is related to #889.

Also, I just now noted that the search-results points to http:// and not http**s**://

**Comment by drybjed** [comment]
Yes, the PR you pointed out changed the structure of the documentation generated on the server. I'll try to wipe the existing build environment on ReadTheDocs and re-build the documentation, perhaps that will help.

**Comment by drybjed** [comment]
Well, I suppose wiping the build environment will not help... From the RTD documentation:

> Remove build environment for master version.
> Having trouble getting your documentation build to complete? By clicking on the button below you'll clean out your build checkouts and environment, but not your generated documentation.

I'm not sure what to do about the search feature. Perhaps RTD has the documentation indexed periodically somehow and just waiting a bit will work?

**Comment by nqb** [comment]
> Perhaps RTD has the documentation indexed periodically somehow and just waiting a bit will work?

Not sure. Currently, if you make a search on "bootstrap" you still have results for `debops.bootstrap` role (which has been removed for a while now). This is certainly why @debuglevel has opened #914.

Moreover, I noticed that [all DebOps builds on RTD](https://readthedocs.org/projects/debops/builds/) are in failed state.

**Comment by drybjed** [comment]
The failed state seems to result from the amount of the documentation in the project. I'm not sure what causes the "assembling single document" step, but that's probably the final generation of the `.html` pages which takes a while.

Perhaps it's time to move the DebOps documentation to our own hosting instead of using ReadTheDocs, where we could have better control over the environment. A `debops.rtfd` role would be needed for this, of course. And I would probably have to setup a new VPS because in the past I got some reports that the current VM the DebOps project uses as a proxy for ReadTheDocs is unreachable from some networks. I guess Hetzner could be a good alternative, I'm not sure at the moment. I'll look at prices first.

**Comment by nqb** [comment]
Sounds like a great idea from my point of view.

**Comment by tumd** [comment]
I have no idea of how ReadTheDocs works and not much experience with GitHub Pages or Gitlab Pages either. But would it be possible to host the doc's on one of the later? Provided that other solutions have higher memory-limits on the builds.

**Comment by drybjed** [comment]
Unfortunately, not. DebOps documentation is written in reStructuredText with Sphinx as the framework. GitHub Pages support content written in Markdown that uses Jekyll to generate static HTML pages. I'm not sure about GitLab Pages but I think that it is similar to GH Pages.

**Comment by nqb** [comment]
Regarding GitLab pages, I'm not sure you can keep different content per branch. I mean you will not be able to have documentation related to `master` branch and for `stable-1.0` at same time.

**Comment by tumd** [comment]
I think Gitlab Pages may fit. GitLab Pages can be generated from what ever backend you choose, as long as the end result of the build is static html, placed in a special pages-folder in the build-environment. And I _think_ the pages-folder persists between builds, which potentially would mean that each built branch could end up in it's own pages/branch-directory.
Here is an (oldish) [example](https://gitlab.com/pages/sphinx).

The runner could even be hosted privately, if the runners provided by GitLab have similar restrictions as RTD. The only other restriction I can think of is if the actual artifact is too large.

**Comment by drybjed** [comment]
> To use this project as your user/group website, you will need one additional step: just rename your project to namespace.gitlab.io, where namespace is your username or groupname.

Well, that's a bummer. That probably would require creating a copy of the `debops/debops` repository as `debops/debops.gitlab.io` and maintaining that, if it would be hosted on GitLab public infrastructure. Private GitLab instance with Pages would probably have a similar requirement, on top of its own stack requirements to make it usable.

I'll probably look into creating the `rtfd` role after `v1.1.0` release, so that we have the option to migrate out from ReadTheDocs.org.

**Comment by alinalexandru** [comment]
@drybjed I think this might be closed.

**Comment by ypid** [comment]
The initial issue is solved, but a task was derived from this. I updated the issue title. I am not sure if we need a `rtfd` role. Using GitLab CI sounds more suitable to me.
